src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -0,0 +1,327 @@
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Core.Imaging;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Adornments;
+using Microsoft.VisualStudio.Text.Editor;
+using AsyncCompletionData = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.AsyncCompletion
+{
+    internal class CompletionSource : ForegroundThreadAffinitizedObject, IAsyncCompletionSource
+    {
+        internal const string RoslynItem = nameof(RoslynItem);
+        internal const string TriggerSnapshot = nameof(TriggerSnapshot);
+        internal const string InsertionText = nameof(InsertionText);
+        internal const string HasSuggestionItemOptions = nameof(HasSuggestionItemOptions);
+        internal const string Description = nameof(Description);
+        internal const string InitialTriggerKind = nameof(InitialTriggerKind);
+        internal const string PotentialCommitCharacters = nameof(PotentialCommitCharacters);
+        internal const string ExcludedCommitCharacters = nameof(ExcludedCommitCharacters);
+        internal const string NonBlockingCompletion = nameof(NonBlockingCompletion);
+
+        private static readonly ImmutableArray<ImageElement> s_WarningImageAttributeImagesArray =
+            ImmutableArray.Create(new ImageElement(Glyph.CompletionWarning.GetImageId(), EditorFeaturesResources.Warning_image_element_automation_name));
+
+        private static readonly EditorOptionKey<bool> NonBlockingCompletionEditorOption = new EditorOptionKey<bool>(NonBlockingCompletion);
+
+        private readonly ITextView _textView;
+
+        internal CompletionSource(ITextView textView, IThreadingContext threadingContext) : base(threadingContext)
+        {
+            _textView = textView;
+        }
+
+        public AsyncCompletionData.CompletionStartData InitializeCompletion(
+            AsyncCompletionData.CompletionTrigger trigger,
+            SnapshotPoint triggerLocation,
+            CancellationToken cancellationToken)
+        {
+            AssertIsForeground();
```

Comment:
Originally, we didn't have them at all. Then, were asked to support. I do not mind to remove assertions at all. Let us just agree.

---
In reply to: [232848833](https://github.com/dotnet/roslyn/pull/29016#discussion_r232848833) [](ancestors = 232848833)

Reply:
An assertion is way of you making a staement about the expectations a method has.  So, for example, if you have a need for a method to run on the UI thread *and* you feel you've controlled everything up till that point to ensure that it *will* definitely be on the UI thread, then you add the assert.  It's something that docs your expectation and fires if violated.

Similarly, if you can be called on any thread, we also have an no-op-assertion for that.  The benefit of this 'assertion' is simply for documentation purposes for the next person that comes along.  This way they don't have to say "hrmm... si this supposed to run only on the ui thread?  only on a BG thread?  both?? i can't tell!".

Reply:
Thank you, @cyrusnajmabadi!
That is what I did. However, I made assertions for both Foreground and Background calls, sync and async. 

The made purpose was to document assumptions.

Then, there were comments:
1. Do not make asserts in async methods.
2. Why do we need to make asserts in CommitManager.ShouldCommitCompletion?

After that, I lost understanding where should we make those checks.

@cyrusnajmabadi, @sharwell, please advise.

---
In reply to: [232900264](https://github.com/dotnet/roslyn/pull/29016#discussion_r232900264) [](ancestors = 232900264)

Reply:
So, again, you should put in the assert if you are making a strong claim that this is a *Requirement* of this method. i.e. if it would be a bug somewhere for someone to call this method on the BG thread, we should put in an 'assertisforeground'.  Similarly, if it's a bug to be called on the UI thread, then we shoudl hve an 'assertisbackground'.  if it's fine to be called on any thread, it should have an 'CanBeCalledOnAnyThread' call in it.  This is most important for sync methods.

The complaints about the asserts before was that the asserts were *wrong*.  Specifically, we were putting in assertisforeground in methods that should not require being called on the FG.  Async mehods, in particular, should be written to be callable by any thread thread (occasionally there are good reasons thsi can't be). If they then *need* to jump to the FG, they should make the appropriate jump (though, really, only if they're JTF ready, as this is super dangerous otherwise).  

So, i would boil it down to:

1. your async methods should likely not require FG/BG.  If they have an assert, it should be 'CanBeCalledOnAnyThread'.  Then, they bounce to whatever thread is appropriate for them.  i.e. GetDesciption should bounce to the BG since it can be super expensive. 
2. your sync methods *in complex multithreading code* should have asserts about if they should be called on fg, bg, or any thread.  This documents our exepctations, and throws when we get things wrong.

Reply:
Thank you very much, @cyrusnajmabadi and @sharwell, for explanations!

Then, I think that we need to keep asserts in the following two cases only:
1. InitializeCompletion
2. TryCommit

The reason for this: we work with SourceText and Snapshots and need to be sure that we exclusively process them. I'll keep asserts there and add some comments. Please correct me if I'm wrong.

---
In reply to: [233199210](https://github.com/dotnet/roslyn/pull/29016#discussion_r233199210) [](ancestors = 233199210)

Reply:
> The reason for this: we work with SourceText and Snapshots and need to be sure that we exclusively process them

What do you mean by 'exclusively process them'?

Reply:
In InitializeCompletion, we want to provide an applicableToSpan. For the purpose, we want locations, positions, and buffers to be frozen during those calculations.

Sorry for confusion! When I mentioned 'process' I referred 'TryCommit'. In InitializeCompletion, we need to make calculations over some object that could change.

---
In reply to: [233661948](https://github.com/dotnet/roslyn/pull/29016#discussion_r233661948) [](ancestors = 233661948)

Reply:
> For the purpose, we want locations, positions, and buffers to be frozen during those calculations.

Gotcha.  However, things like SourceTexts ans Snapshots (and SnapshotPoints) are naturally frozen by design.  They're safe to use in the BG as long as you captured them at a consistent point and pass htem along as is.

Reply:
Thank you!
We want to provide an ApplicableToSpan to the editor in the InitializeCompletion. Let us be sure that nobody changes anything that can break our calculation. If we need to support this in the BG, let us have a separate discussion what scenarios require this and how should we sync this with the FG.

---
In reply to: [233709305](https://github.com/dotnet/roslyn/pull/29016#discussion_r233709305) [](ancestors = 233709305)

Reply:
Issues with assertions should be resolved now.

---
In reply to: [232883547](https://github.com/dotnet/roslyn/pull/29016#discussion_r232883547) [](ancestors = 232883547,232848833)